### PR TITLE
Integration specs

### PR DESCRIPTION
**Pull request by AlierPM**
# Blog App - Integration Specs for Views and Fix N+1 Problems

## Learning Objectives

- Wrote integration tests with the Capybara gem.
- Secured the app from N+1 problems.

## Time Spent

Approximately 4.5 hours.

## Description

In this project, integration tests were created for all views in the app. User stories or workflows were covered, considering possible errors users may cause.

**IMPORTANT NOTE:** Please review all requirements to ensure completion.

## General Requirements

- Ensured there are no linter errors.
- Used correct Gitflow.
- Documented work professionally.

## Ruby Requirements

- Followed best practices for Ruby.

### N+1

- Solved the N+1 problem when fetching all posts and their comments for a user.
- Did not use the bullet gem in this exercise; analyzed the logs as shown in [this article](link-to-article).
- Included screenshots of the app's logs before and after the fix in the PR description.
Screenshot from 2024-01-26 21-07-45

Screenshot from 2024-01-26 21-20-04

### Integration Specs

Used Capybara to write integration tests for each view in the project.

#### User Index Page

- Could see the username of all other users.
- Could see the profile picture for each user.
- Could see the number of posts each user has written.
- Clicking on a user redirected to that user's show page.

#### User Show Page

- Could see the user's profile picture.
- Could see the user's username.
- Could see the number of posts the user has written.
- Could see the user's bio.
- Could see the user's first 3 posts.
- Could see a button that lets me view all of a user's posts.
- Clicking a user's post redirected to that post's show page.
- Clicking to see all posts redirected to the user's post index page.

#### User Post Index Page

- Could see the user's profile picture.
- Could see the user's username.
- Could see the number of posts the user has written.
- Could see a post's title.
- Could see some of the post's body.
- Could see the first comments on a post.
- Could see how many comments a post has.
- Could see how many likes a post has.
- Could see a section for pagination if there are more posts than fit on the view.
- Clicking on a post redirected to that post's show page.

#### Post Show Page

- Could see the post's title.
- Could see who wrote the post.
- Could see how many comments it has.
- Could see how many likes it has.
- Could see the post body.
- Could see the username of each commentor.
- Could see the comment each commentor left.

### CODE REVIEW

File: spec/features/post_show_spec.rb [hunk 1]
```diff
@@ -32,7 +32,6 @@
     create(:comment, post:, user: commentor, text: 'Nice post!')
```
Comment (thecodechaser) [inline spec/features/post_show_spec.rb:33]: - [x] All of the tests are failing, Please make sure that all tests pass without any errors.

image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @AlierPM, and @Eccb7👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️
- Giflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/post_show_spec.rb (hunk 1 above)
Comment (AlierPM) [reply spec/features/post_show_spec.rb:33]: @thecodechaser; It was Tailwind CSS file path conflict. It is now resolved and Now all the tests are passing. Thank for the constructive reviews.

**Comment by AlierPM** [comment]
Screenshot from 2024-01-30 19-44-59

**Comment by shahadat3669** [comment]
@AlierPM, Well! while I was reviewing your project! I noticed that you pushed a new commit with some changes. it is not good practice to push new changes after requesting a review. it became hard for the reviewer and has to review again with the new changes. so I will recommend you update your code as much as possible before the code review request and after requesting a review wait for feedback.

**Comment by AlierPM** [comment]
@shahadat3669; well noticed!

**Review by shahadat3669 — APPROVED** [review]
Hi @AlierPM, @Eccb7, 

Your project is complete!  There is nothing else to say other than it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✅ No linter Errors
✅  All test cases passed
✅ The code is top-notch! 💯

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**